shell/platform/android/io/flutter/view/AccessibilityViewEmbedder.java

```diff
@@ -220,6 +225,122 @@ private void copyAccessibilityFields(AccessibilityNodeInfo input, AccessibilityN
         }
     }
 
+    /**
+     * Delegates an AccessibilityNodeProvider#requestSendAccessibilityEvent from the AccessibilityBridge to the embedded
+     * view.
+     *
+     * @return True if the event was sent.
+     */
+    public boolean requestSendAccessibilityEvent(View embeddedView, View eventOrigin, AccessibilityEvent event) {
+        AccessibilityEvent translatedEvent = AccessibilityEvent.obtain(event);
+        Long originPackedId = reflectionAccessors.getRecordSourceNodeId(event);
+        if (originPackedId == null) {
+            return false;
+        }
+        int originVirtualId = ReflectionAccessors.getVirtualNodeId(originPackedId);
+        Integer flutterId = originToFlutterId.get(new ViewAndId(embeddedView, originVirtualId));
+        if (flutterId == null) {
+            return false;
+        }
+        translatedEvent.setSource(rootAccessibilityView, flutterId);
+        translatedEvent.setClassName(event.getClassName());
+        translatedEvent.setPackageName(event.getPackageName());
+
+        for (int i = 0; i < translatedEvent.getRecordCount(); i++) {
+            AccessibilityRecord record = translatedEvent.getRecord(i);
+            Long recordOriginPackedId = reflectionAccessors.getRecordSourceNodeId(record);
+            if (recordOriginPackedId == null) {
+                return false;
+            }
+            int recordOriginVirtualID = ReflectionAccessors.getVirtualNodeId(recordOriginPackedId);
+            Integer recordFlutterId = originToFlutterId.get(new ViewAndId(embeddedView, recordOriginVirtualID));
+            if (recordFlutterId == null) {
+                return false;
+            }
+            record.setSource(rootAccessibilityView, recordFlutterId);
+        }
+
+        return rootAccessibilityView.getParent().requestSendAccessibilityEvent(eventOrigin, translatedEvent);
+    }
+
+    /**
+     * Delegates an @{link AccessibilityNodeProvider#performAction} from the AccessibilityBridge to the embedded view's
+     * accessibility node provider.
+     *
+     * @return True if the action was performed.
+     */
+    public boolean performAction(int flutterId, int accessibilityAction, Bundle arguments) {
+        ViewAndId origin  = flutterIdToOrigin.get(flutterId);
+        if (origin == null) {
+            return false;
+        }
+        View embeddedView = origin.view;
+        AccessibilityNodeProvider provider = embeddedView.getAccessibilityNodeProvider();
+        if (provider == null) {
+            return false;
+        }
+        return provider.performAction(origin.id, accessibilityAction, arguments);
+    }
+
+    /**
+     * Returns a flutterID for an accessibility record, or null if no mapping exists.
+     *
+     * @param embeddedView the embedded view that the record is associated with.
+     */
+    public Integer getRecordFlutterId(View embeddedView, AccessibilityRecord record) {
```

Comment:
I'd either tag this method with `@Nullable` and document that it returns null when one can't be found or convert it to use `int` with an `INVALID` constant for not found cases. It's subjective but I prefer the `INVAILD` constant option instead since it's more clear and can't potentially end up with NPEs.

Reply:
I feel like returning a valid int value here is dangerous as misuse might result in an error that's much harder to spot (setting focus on a wrong a11y node, or setting a wrong child node) vs. an NPE.

Reply:
I see the reasoning. I think adding the `@Nullable` annotations helps with the discoverability here at any rate, so at least now callers of the method have it declared for them.